src/ddspdrum/parameter.py

```diff
@@ -0,0 +1,159 @@
+        value (float)   : value within [0,1] range to convert to range defined by
+            minimum and maximum
+        """
+        assert torch.all(0.0 <= value)
+        assert torch.all(value <= 1.0)
+
+        if self.curve != 1:
+            value = torch.exp2(torch.log2(value) / self.curve)
+
+        return self.minimum + (self.maximum - self.minimum) * value
+
+    def to_0to1(self, value: T) -> T:
+        """
+        Convert a ranged parameter to a normalized range from 0 to 1
+
+        Parameters
+        ----------
+        value (float)   : value within the range defined by minimum and maximum
+        """
+        assert torch.all(self.minimum <= value)
+        assert torch.all(value <= self.maximum)
+
+        normalized = (value - self.minimum) / (self.maximum - self.minimum)
+        if self.curve != 1:
+            normalized = torch.pow(normalized, self.curve)
+
+        return normalized
+
+
+class TorchParameter(nn.Parameter):
+    """
+    Parameter class that inherits from pytorch nn.Parameter so it can be used for
+    training. Can use a ParameterRange object to help convert from a 0 to 1 range
+    which is expected internally and an external user specified range.
+
+    Parameters
+    ----------
+    value (float) : initial value of this parameter in the user-specific range.
+        Must pass in a ParameterRange object when using this to provide conversion to
+        and from 0-to-1 range
+    parameter_name (str) : A name for this parameter
+    parameter_range (ParameterRange) : A ParameterRange object that supports conversion
+        to and from 0-to-1 range and a user-specified range.
+    data (Tensor) : directly add data to this parameter without a user-range
+    requires_grad (bool) : whether or not a gradient is required for this parameter
+    """
+
+    def __new__(
+            cls,
+            value: float = None,
+            parameter_name: str = "",
+            parameter_range: ParameterRange = None,
+            data: torch.Tensor = None,
+            requires_grad: bool = True,
+    ):
+        if value is not None:
+            if parameter_range is not None:
+                data = parameter_range.to_0to1(T(value))
+            else:
+                raise ValueError(
+                    "A parameter range must be specified when passing in a value"
+                )
+
+        self = super().__new__(cls, data, requires_grad)
+
+        # Additional members -- check to make sure they don't exist first
+        # (This is sanity check in case something changes in pytorch in the future)
+        assert 'parameter_range' not in self.__dict__
+        self.parameter_range = parameter_range
+
+        assert 'parameter_name' not in self.__dict__
+        self.parameter_name = parameter_name
+
+        return self
+
+    def __repr__(self):
+        return "TorchParameter(name={}, value={})".format(
+            self.parameter_name, self.item()
+        )
+
+    def get_in_range(self) -> T:
```

Comment:
Do we want to call this from_0to1 and to_0to1? I just find it confusing each to time think through what "in range" means because it's kinda idiomatic and could mean either convention